### PR TITLE
V2: Add pytest for CLI functionality

### PR DESCRIPTION
**Pull request by hwpang**
## Description
This PR addresses https://github.com/chemprop/chemprop/issues/590. I added pytest for the CLI functionality. 

## Questions
- Currently the test only tests that the CLI can run and use only 1 epoch. Should we increase the epoch and test the model performance? For example, we can pass the model generated by the training CLI test, and load it using the predicting CLI, and test the errors. I don't think it's necessary as this test focuses on the CLI connection aspect, and the model performance is tested by the integration test. But I want to know people's thoughts.

## Relevant issues
This PR addresses https://github.com/chemprop/chemprop/issues/590.

### CODE REVIEW

File: tests/cli/test_cli_regression_mol+mol.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+def data_path(data_dir):
+    return str(data_dir / "regression" / "mol+mol.csv")
+
+
+@pytest.fixture
+def checkpoint_path(data_dir):
+    return str(data_dir / "example_model_v2_regression_multi.ckpt")
+
+
+def test_training(monkeypatch, data_path):
```
Comment (davidegraff) [inline tests/cli/test_cli_regression_mol+mol.py:21]:
```suggestion
def test_train(monkeypatch, data_path):
```

File: tests/cli/test_cli_regression_mol+mol.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    args = ["chemprop", "train", "-i", data_path, "--smiles-columns", "smiles", "solvent", "--epochs", "1", "--num-workers", "4"]
+
+    with monkeypatch.context() as m:
+        m.setattr("sys.argv", args)
+        main()
+
+
+def test_predicting(monkeypatch, data_path, checkpoint_path):
```
Comment (davidegraff) [inline tests/cli/test_cli_regression_mol+mol.py:29]:
```suggestion
def test_predict(monkeypatch, data_path, checkpoint_path):
```

File: tests/cli/test_cli_regression_mol+mol.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return str(data_dir / "regression" / "mol+mol.csv")
+
+
+@pytest.fixture
+def checkpoint_path(data_dir):
+    return str(data_dir / "example_model_v2_regression_multi.ckpt")
+
+
+def test_training(monkeypatch, data_path):
+    args = ["chemprop", "train", "-i", data_path, "--smiles-columns", "smiles", "solvent", "--epochs", "1", "--num-workers", "4"]
```
Comment (davidegraff) [inline tests/cli/test_cli_regression_mol+mol.py:22]: we can't really assume we'll have access to more than 1 CPU
```suggestion
    args = ["chemprop", "train", "-i", data_path, "--smiles-columns", "smiles", "solvent", "--epochs", "1", "--num-workers", "0"]
```

**Review by davidegraff — COMMENTED** [review]
I think this is a good first start to testing the CLI. To your question of whether we want to test that the CLI actually trains the model and that we can use that model for downstream prediction, I don't think we do. The core of this question is: "What are we trying to test?" I would argue that the CLI is only responsible for structuring inputs corresponding to our arguments and that training is handled at the module level (and thus already tested, as you've mentioned). So we should really be testing a few things:
1. that it runs (similar to `test_quick` in our integration tests)
2. that it produces the same output structure we say it should (based on where the conversation of #591 ends up)
3. that the actual outputs mirror expectations:
    - we produce the expected model architecture
    - our splits are correct (if that can be checked)
    - the command config is correct
 
I don't think that all of these need to be addressed here, but they should guide our future work on the CLI testing suite.

Lastly, I didn't know about `monkeypatch`--good find! Your solution is much better than what I originally proposed!

**Comment by hwpang** [comment]
Adding some discussions from today's chemprop dev 2 meeting
- Add as many arguments that are available for training and predicting to the tests as possible
- Train more epochs and compare the model results to catch errors such as someone accidentally delete the line to feed `final_lr` in.

**Comment by hwpang** [comment]
@davidegraff Thanks for the comments! I made the following updates according to your comments.

1. I updated the first test to `test_train_quick` and `test_predict_quick` test that `chemprop train` and `chemprop predict` works.
2. Regarding the output directory structure, I am currently testing that the expected output directories exist. I'm happy to take suggestions on better way to test these. Additionally, I wrote the tests according to the current output directory structure as the discussions in https://github.com/chemprop/chemprop/issues/591 and the changes in https://github.com/chemprop/chemprop/pull/599 have not been merged in yet.
3. Regarding testing the model architecture, I am currently testing that the model checkpoint file can be loaded. I'm happy to take suggestions on a better way to test this.

**Review by davidegraff — APPROVED** [review]

**Pull request merged**